src/Elrond/Serialization.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright © 2017-2020 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Serialization.h"
+
+#include "../Elrond/Address.h"
+#include "../proto/Elrond.pb.h"
+#include "Base64.h"
+#include "PrivateKey.h"
+
+using namespace TW::Elrond;
+
+std::map<string, int> fields_order {
+    {"nonce", 1},
+    {"value", 2},
+    {"receiver", 3},
+    {"sender", 4},
+    {"gasPrice", 5},
+    {"gasLimit", 6},
+    {"data", 7},
+    {"signature", 8}
+};
+
+struct FieldsSorter {
+    bool operator() (const string& lhs, const string& rhs) const {
+        return fields_order[lhs] < fields_order[rhs]; 
+    }
+};
+
+template<class Key, class T, class Compare, class Allocator>
+using sorted_map = std::map<Key, T, FieldsSorter, Allocator>;
+using sorted_json = nlohmann::basic_json<sorted_map>;
+
+string TW::Elrond::serializeTransactionToSignableString(const Proto::TransactionMessage& message) {
+    sorted_json payload {
+        {"nonce", json(message.nonce())},
+        {"value", json(message.value())},
+        {"receiver", json(message.receiver())},
+        {"sender", json(message.sender())},
+        {"gasPrice", json(message.gas_price())},
+        {"gasLimit", json(message.gas_limit())},
+    };
+
+    if (!message.data().empty()) {
+        payload["data"] = json(TW::Base64::encode(TW::data(message.data())));
+    }
+
+    return payload.dump();
+}
+
+string TW::Elrond::serializeSignedTransaction(const Proto::TransactionMessage& message, string signature) {
+    sorted_json payload {
+        {"nonce", json(message.nonce())},
+        {"value", json(message.value())},
+        {"receiver", json(message.receiver())},
+        {"sender", json(message.sender())},
+        {"gasPrice", json(message.gas_price())},
+        {"gasLimit", json(message.gas_limit())},
+        {"data", json(message.data())},
```

Comment:
looks duplicate code in `serializeTransactionToSignableString`, also, `data` is not checked?

Reply:
There is duplication, indeed - a bit closer to _accidental_ duplication though, and we would like to keep `serializeTransaction` (called `serializeTransactionToSignableString` before review) separately.

Indeed, the `data` field is applied a special treatment in the first serialization method, whose output is meant to be signed - and we have a more constrained procedure here, such as: omit the data field if empty, apply _base64_ on it. On the other hand, these constraints do not apply to obtaining the actual message that should be dispatched to the blockchain.

Is it acceptable on your side if we keep these as they are?